sql/src/main/java/io/crate/operation/join/BitSet.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation.join;
+
+import org.apache.lucene.util.LongBitSet;
+
+import java.util.ArrayList;
+
+/**
+ * This BitSet is used to mark matched rows between left and right in {@link NestedLoopOperation}
+ *
+ * Each bit true if the rows in the respective position are matched and therefore we need
+ * a structure capable of holding <pre>long</pre> size of bits so java.util.BitSet cannot be used.
+ *
+ * We choose to use {@link LongBitSet} from Lucence and add another layer of segments on top in
+ * order to further optimize performance. The usage of segments prevents from calling System.arrayCopy()
+ * and instead when the current segments are full we reserve a new segment and add it to the segments
+ * list.
+ */
+class BitSet {
+
+    private static long SEGMENT_SIZE = 1048576;
+    private static int INITIAL_NO_OF_SEGMENTS = 10;
```

Comment:
FixedBitSet has get(int index) but we discussed with Bernd that we want long to be able to manage large data sets on the right side. Furthermore we need to be able to dynamically grow since we don't know the size beforehand. The purpose of starting with 10 segments of 1024^2 bits is not to reserve too much memory in the beginning but still be able to grow dynamically. We could directly use the LongBitSet and the ensureCapacity but this makes a copy of all the existing bits in order to grow, where in our case we just grow the List of segments which holds references to the fixed underlying LongBitSets.